### DISCUSSION TITLE
How to include pinocchio library in a project as a header only library?

### DISCUSSION THREAD

**Discussion by IvanVnucec** [stack-of-tasks/pinocchio#1673] [Q&A]
How to include pinocchio library in a c++ project as a header only library? Is it possible?

**Comment by igor-sadalski** [comment]
I am trying to use CMake

**Comment by jcarpent** [comment]
Dear @IvanVnucec and @igor-sadalski,

The answer is yes, but it is not as easy as we would like.
The core framework is header-only while only the URDF and Python parsers are shared libs.

@IvanVnucec Why do you specifically need header-only?

**Reply by jcarpent** [reply]
We can also think to export an extra CMake target which will correspond to the header-only part of Pinocchio

**Reply by IvanVnucec** [reply]
Basically I am trying to simulate flying quadcopter. Quadcopter model is really simple one with two rods crossed with a fixed joint. On every rods end there is a propeller simulated as a simple force. 
Simulator is written in C++ and I'm not intending to use Pinocchio Python API.
As for URDF, my quadcopter model is so simple that I think I could write it by hand in C++ if possible.

Is it possible that I could incorporate Pinocchio library as a header only library? If yes, is there any documentation explaining how?

**Reply by jcarpent** [reply]
If you don't use the parsers, then you will just need to link to the headers of Pinocchio.
I will try to do a demo by Monday. Do you use CMake?

**Reply by IvanVnucec** [reply]
Great, thank you! Yes, I am using CMake.

**Reply by fabinsch** [reply]
Hey @IvanVnucec and @igor-sadalski , I opened a PR to add an additional target called `pinocchio_headers` in the CMakeLists - you can check what I have done here #1712.

Now I can compile this example file (test_forward_dynamics.cpp):

```
#include "pinocchio/parsers/sample-models.hpp"

#include "pinocchio/algorithm/joint-configuration.hpp"
#include "pinocchio/algorithm/aba.hpp"

#include <iostream>

int main(int argc, char ** argv)
{
  using namespace pinocchio;
  
  pinocchio::Model model;
  pinocchio::buildModels::humanoidRandom(model);
  
  // Build a data related to model
  Data data(model);
  
  // Sample a random joint configuration as well as random joint velocity and torque
  Eigen::VectorXd q = randomConfiguration(model);
  Eigen::VectorXd v = Eigen::VectorXd::Zero(model.nv);
  Eigen::VectorXd tau = Eigen::VectorXd::Random(model.nv);
  
  // Computes the forward dynamics (ABA) 
  aba(model, data, q, v, tau);
  
  // Get access to the joint acceleration
  std::cout << "Joint acceleration: " << data.ddq << std::endl;
}
```
with the command
```
c++ -o test test_forward_dynamics.cpp -L<path-to-pinocchio_headers_lib> -lpinocchio_headers -I$CONDA_PREFIX/include/eigen3 -I$CONDA_PREFIX/include/
```
You need to provide the path to eigen3 and boost (in my case in a conda env), as well as the pinocchio_headers.

**Reply by IvanVnucec** [reply]
Great, very nicely written. Thank you for commitment!
